solr/core/src/java/org/apache/solr/handler/admin/api/CoreAdminAPIBase.java

```diff
@@ -0,0 +1,210 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.solr.handler.admin.api;
+
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.concurrent.ExecutorService;
+import java.util.function.Supplier;
+import org.apache.solr.api.JerseyResource;
+import org.apache.solr.common.SolrException;
+import org.apache.solr.common.util.ExecutorUtil;
+import org.apache.solr.common.util.SolrNamedThreadFactory;
+import org.apache.solr.core.CoreContainer;
+import org.apache.solr.handler.api.V2ApiUtils;
+import org.apache.solr.jersey.SolrJerseyResponse;
+import org.apache.solr.logging.MDCLoggingContext;
+import org.apache.solr.request.SolrQueryRequest;
+import org.apache.solr.response.SolrQueryResponse;
+import org.apache.solr.util.tracing.TraceUtils;
+import org.slf4j.MDC;
+
+public abstract class CoreAdminAPIBase extends JerseyResource {
+
+  private static final ExecutorService PARALLEL_EXECUTOR =
```

Comment:
[-1] I love the approach on this class overall (esp. the `handle` method that takes in a Supplier), but I'm concerned about reliance on static state here.

I hinted at this a bit on JIRA, but should've been more explicit:

> I guess we could create a base-class for our APIs, e.g. "AsyncTrackingApiBase", with a static instance to hold this status info, that gets extended by various API subclasses. I don't love the idea of introducing non-constant static state, but I think it'd work?

It's my fault for not being more explicit: it's not the norm, but it's also not unheard of to run multiple Solr instances from within the same JVM.  Solr's tests do it, for one, and I've heard of actual applications doing this too on occasion via EmbeddedSolrServer and similar classes.

Maybe this would "just work" if the tests and EmbeddedSolrServer know to use different classloaders, but if it doesn't, then the use of static state could cause tricky bugs - e.g. the same executor being used for multiple Solr instances.

I'll run the test suite and poke around some of the related code to double-check my fears here, but assuming they're well founded, we'll need a different approach here.

One approach that's only slightly a "dodge" would be to essentially re-use the same status-map and `parallelExecutor` that CoreAdminHandler creates.  (i.e. Add getters for these things to `CoreAdminHandler` and then modify the Jersey application to make the status-map and executor injectable as singletons, similar to what we do for CoreContainer [here](https://github.com/apache/solr/blob/c86128855807042e25c5bd320e6f7729784e8884/solr/core/src/java/org/apache/solr/core/CoreContainer.java#L1067))

This would have a few side-benefits:

- avoids the creation of an additional executor (the current PR creates one in the handler and one in this API class)
- no executor lifecycle changes required, since we're reusing the executor
- injection gives the API classes enough indirection that they won't require any changes when we eventually nuke v1 and need to find a different way of managing the executor/status-map
- no duplication of constants and status-map init logic (current PR has this code in both the handler and this API class).

Any thoughts on that?  I'm happy to make the changes required as the rework is largely my fault for not communicating my concerns around static-state strongly enough.

Reply:
Oh gotcha, not a problem - The way I read your original comment, I thought we were going fully with the static approach first, I can take some blame for my interpretation. I think what you're laying out here makes sense and I hadn't considered the case of multiple Solr Contexts running on a single VM. I took a little bit of time to see how I would register the map/executor, but haven't made any code changes yet. I will attempt to work on it this weekend if you haven't already gotten to it.

Would migrating the status map and executor into a class that can then be used by both the CoreAdminHandler and made injectable for the CoreAdminAPIBase work? That may save us from having to inject multiple things and could potentially get us closer to the future state of what async could look like.